### PR TITLE
ASoC: SOF: Intel: hda_pcm: Regulate DPIB reading at frame boundary for playback stream

### PR DESCRIPTION
**Pull request by yongzhi1**
ChromeOS alsa_conformance_test [1] uses linear regression to define
the relationship between time elapsed and frame played. The measured
slope and variance are two criteria used to determine pass/fail
of the sample rate tests, RATE_CRITERIA_DIFF_PCT and
RATE_ERR_CRITERIA are defaulted to 0.01 and 10 respectively.

The test failed randomly for either rate or mean squared residuals on JSL+
platform, repro rate close to 100%, and with 20us applied, the same test
pass consistently on Waddledoo and Drawlat.

[1] https://chromium.googlesource.com/chromiumos/platform/audiotest/+/HEAD/alsa_conformance_test.md#Stability-of-rate
[2] Use "alsa_conformance_test -P hw:0,0 -f S24_LE --debug" to view the sample data.

Signed-off-by: Yong Zhi <yong.zhi@intel.com>

### CODE REVIEW

**Comment by yongzhi1** [comment]
Documented the rate test:
https://chromium-review.googlesource.com/c/chromiumos/platform/audiotest/+/2372962

Sample output:
[alsa_conformance_test-S24-results.txt](https://github.com/thesofproject/linux/files/5119146/alsa_conformance_test-S24-results.txt)

With the python script:

localhost ~ # alsa_conformance_test.py -P hw:0,0                                                     
7 passed, 2 failed                                                                                   
Device Information                                                                                   
        Name: hw:0,0                                                                                 
        Card: sofda7219max983 [sof-da7219max98360a]                                                  
        Device: Speakers (*) []                                                                      
        Stream: PLAYBACK                                                                             
        Format: ['S16_LE', 'S24_LE', 'S32_LE']                                                       
        Channels: [2]                                                                                
        Rate: [48000]                                                                                
        Period_size range: [24, 4080]                                                                
        Buffer_size range: [48, 16368]                                                               
Test Params                                                                                          
        Set channels 2: pass                                                                         
        Set format S16_LE: pass                                                                      
        Set format S24_LE: pass                                                                      
        Set format S32_LE: pass                                                                      
        Set rate 48000: pass                                                                         
Test Rates                                                                                           
        Set rate 48000: fail - Expected rate is 48000, measure 48007.453312999998, difference 7.45338
Test All Pairs                                                                                       
        Set channels 2, format S16_LE, rate 48000: fail - Expected rate is 48000, measure 48007.73118
        Set channels 2, format S24_LE, rate 48000: pass                                              
        Set channels 2, format S32_LE, rate 48000: pass

localhost ~ # alsa_conformance_test.py -P hw:0,0
8 passed, 1 failed
Device Information
        Name: hw:0,0
        Card: sofda7219max [sof-da7219max]
        Device: Speakers  []
        Stream: PLAYBACK
        Format: ['S16_LE', 'S24_LE', 'S32_LE']
        Channels: [2]
        Rate: [48000]
        Period_size range: [24, 4080]
        Buffer_size range: [48, 16368]
Test Params
        Set channels 2: pass
        Set format S16_LE: pass
        Set format S24_LE: pass
        Set format S32_LE: pass
        Set rate 48000: pass
Test Rates
        Set rate 48000: pass
Test All Pairs
        Set channels 2, format S16_LE, rate 48000: fail - Rate error 10.929639999999999 > threshold 10
        Set channels 2, format S24_LE, rate 48000: pass
        Set channels 2, format S32_LE, rate 48000: pass

**Comment by yongzhi1** [comment]
> this change makes no sense to me, sorry.
> the hw_ptr is tied to the DMA pointer, which will increment in bursts when the firmware enables new requests, e.g. it will typically progress in chunks of 1ms.
> If you want to relate playback time and frame played, then you need to use snd_pcm_delay().
> 
> I don't think we should go ahead with this.

Hi @plbossart , thanks for your comments, here is the main logic of the playback, do you see any issue of the test itself?

https://chromium.googlesource.com/chromiumos/platform/audiotest/+/refs/heads/master/alsa_conformance_test/alsa_conformance_thread.c#306

https://chromium.googlesource.com/chromiumos/platform/audiotest/+/refs/heads/master/alsa_conformance_test/alsa_conformance_helper.c#460

**Review by kv2019i — CHANGES_REQUESTED** [review]
@yongzhi1 Do you have debug dumps of a failing case to share? @plbossart The test is basicly busylooping on snd_pcm_avail() and storing a data point whenever the value changes. It's not really interested in playback delay at all, so I don't think snd_pcm_delay() is needed. The test would seem to be correct, as even with our bursty DMA, over longer test sequence ("long term" in the test tool), the rate of consumption should match the sampling rate. 

I agree with @plbossart the patch in its current form (and I'd argue the original workaround for capture patch smells pretty fishy as well, but that seems to have been there for a long time and legacy driver has adopted the same trick). Adding the usleep to the playback path might be helping just by slowing down the busylooping logic. There's still something odd happening here -- maybe the DPIB based position query returns an occasional crazy value and the linear regression picks this up.

**Comment by yongzhi1** [comment]
> @yongzhi1 Do you have debug dumps of a failing case to share? @plbossart The test is basicly busylooping on snd_pcm_avail() and storing a data point whenever the value changes. It's not really interested in playback delay at all, so I don't think snd_pcm_delay() is needed. The test would seem to be correct, as even with our bursty DMA, over longer test sequence ("long term" in the test tool), the rate of consumption should match the sampling rate.
> 
> I agree with @plbossart the patch in its current form (and I'd argue the original workaround for capture patch smells pretty fishy as well, but that seems to have been there for a long time and legacy driver has adopted the same trick). Adding the usleep to the playback path might be helping just by slowing down the busylooping logic. There's still something odd happening here -- maybe the DPIB based position query returns an occasional crazy value and the linear regression picks this up.

Hi. @kv2019i and @plbossart, 

Thanks for the review, assuming:
1. Test method is valid
2. No real issue with sample rate performance for jsl+ platform.

Erroneous hw_ptr update is very likely the cause for test failure, of cause, this patch is merely based on test results , either a better fix or a theory is needed to make sense of it.

For the debug dump, there is no error/warning in the kernel traces, the test setup is very simple [wiki](https://chromium.googlesource.com/chromiumos/platform/audiotest/+/refs/heads/master/alsa_conformance_test.md).
For a typical test dump: [alsa_conformance_test-S24-results.txt](https://github.com/thesofproject/linux/files/5119146/alsa_conformance_test-S24-results.txt)

Notice, I have purposely removed the "outliers" during the debug, the test results get improved slightly but I am still not able to get rid of the error, nothing like the 20us magic.

**Comment by plbossart** [comment]
@yongzhi1 I am 100% positive that if you want to look at playback rate, you *NEED* to look at the delay information. Imagine if there's a 2s buffer in the DSP, looking at the DMA information would be taking into account very large steps. This test was made assuming there's no buffering I guess.

**Comment by kv2019i** [comment]
@yongzhi1 wrote:
>     1. Test method is valid
>     2. No real issue with sample rate performance for jsl+ platform.

Hmm, not sure about (1) anymore. The linear regression is done on amount of samples consumed and time it took, but for any individual update, this time is related to DMA speed. The merging logic does hide this, but in the log you share, the merging does not pick up all cases:

normal case, 48 samples consumed at expected rate
0.000999428          432      18768         48       48027.471714

merge case
0.000999810          408      18792         24       24004.560867
0.000026097          384      18816         24      919645.936315 [Merged]

no merge case
0.000937183          380      11380          4        4268.109857
0.000024140          336      11424         44     1822700.911350

... so here as input to linear regression, we capture a record where consumption of 4 samples took 0.937ms, followed by consumption of 44 samples in time of 0.024ms. If these are used directly as inputs to linear regression, this will skew results and show up as a large rate error.

And I think the above readings from driver seem valid. If you sample hw_ptr/avail during a (~1ms) DMA burst, you may see avail to move very rapidly (like above no merge case).

**Comment by yongzhi1** [comment]
> @yongzhi1 wrote:
> 
> > ```
> > 1. Test method is valid
> > 2. No real issue with sample rate performance for jsl+ platform.
> > ```
> 
> Hmm, not sure about (1) anymore. The linear regression is done on amount of samples consumed and time it took, but for any individual update, this time is related to DMA speed. The merging logic does hide this, but in the log you share, the merging does not pick up all cases:
> 
> normal case, 48 samples consumed at expected rate
> 0.000999428 432 18768 48 48027.471714
> 
> merge case
> 0.000999810 408 18792 24 24004.560867
> 0.000026097 384 18816 24 919645.936315 [Merged]
> 
> no merge case
> 0.000937183 380 11380 4 4268.109857
> 0.000024140 336 11424 44 1822700.911350
> 
> ... so here as input to linear regression, we capture a record where consumption of 4 samples took 0.937ms, followed by consumption of 44 samples in time of 0.024ms. If these are used directly as inputs to linear regression, this will skew results and show up as a large rate error.
> 
> And I think the above readings from driver seem valid. If you sample hw_ptr/avail during a (~1ms) DMA burst, you may see avail to move very rapidly (like above no merge case).

Hi @kv2019i ,
Default value for merge_threshold_t is used for the test:
merge_threshold_t: 0.000100 <= this can be set, do you have a recommended value?
merge_threshold_sz: 29 <= this is auto calculated 

Thanks,
Yong

**Comment by yongzhi1** [comment]
> @yongzhi1 I am 100% positive that if you want to look at playback rate, you _NEED_ to look at the delay information. Imagine if there's a 2s buffer in the DSP, looking at the DMA information would be taking into account very large steps. This test was made assuming there's no buffering I guess.

Hi, @plbossart ,

Current test runs 1s with default --block_size set to 240:
step min: 3
step max: 48
step median: 48

step max looks OK right? [stability-of-step](https://chromium.googlesource.com/chromiumos/platform/audiotest/+/refs/heads/master/alsa_conformance_test.md#stability-of-step)

**Comment by kv2019i** [comment]
@yongzhi1 I think if you could force the merge threshold to 48 samples (i.e. size of the DMA burst), the impact of burst timing would be hidden and tests would probably pass. 

But this leaves the question whether the test is really valid in its current form. The DMA burst size is not explicitly declared and discoverable by applications. If you had a system with a big internal buffer (like a 2sec buffer), adjusting the merge threshold would not help.

So that brings us back to point @plbossart made that to measure the playback rate in this granularity, you'd have to factor in snd_pcm_delay(). Or average the rate of data consumption (by DMA) over a longer period of time.

Maybe good to check with the authors. Some of the checks are clearly intentional to meet CRAS reqs, but purely from ALSA interface point of view, driver is not out of spec here.

PS One driver side hack that would probably make the tests pass would be to align position reporting to DMA burst size. But this would limit granularity of our delay reporting as well, so does not seem like a good change overall.

**Comment by yongzhi1** [comment]
Hi @kv2019i and @plbossart ,

Will take your points to do some more experiment with the test, however, I am still bothered that the added 20us should not alter the characteristic of the burst nature of DMA, but the output is rather too dramatic:
[alsa_conformance_test-S24-results-2.txt](https://github.com/thesofproject/linux/files/5125424/alsa_conformance_test-S24-results-2.txt)

**Comment by yongzhi1** [comment]
Hi, @kv2019i,
I made some changes to the test code to set merge_threshold_sz, here is the dump:
[alsa_conformance_test-S16-results.txt](https://github.com/thesofproject/linux/files/5127087/alsa_conformance_test-S16-results.txt)

In the 1st run, merge_threshold_sz 32 is calculated based on step median and rate error: 11.688836.
In the 2nd run, merge_threshold_sz 48 is set from user input, and rate error: 5.039372.

Please let me know if the above behavior agrees with your recommendation. Thanks!!
(test patch https://chromium-review.googlesource.com/c/chromiumos/platform/audiotest/+/2375921)

**Comment by plbossart** [comment]
> V2:
> 
> 1. Updated based on Kar's suggestion to avoid the 20us, it serves the purpose for alsa_complance_test, but not sure whether DPIB has specific advantage when used here.
> 2. I can avoid changes to capture if so preferred, seems the code is deliberated done to fix some issue.

I don't see the point of using LPIB. This was meant to be used when we transmit on a link, but here these are bursty DMA transfers unrelated to the HDaudio framing. In fact the sample alignment *guarantees* a 20us uncertainty on what the DMA did.

The code we have doesn't seem to be broken, so it it ain't broke don't fix it?

**Comment by yongzhi1** [comment]
> > V2:
> > 
> > 1. Updated based on Kar's suggestion to avoid the 20us, it serves the purpose for alsa_complance_test, but not sure whether DPIB has specific advantage when used here.
> > 2. I can avoid changes to capture if so preferred, seems the code is deliberated done to fix some issue.
> 
> I don't see the point of using LPIB. This was meant to be used when we transmit on a link, but here these are bursty DMA transfers unrelated to the HDaudio framing. In fact the sample alignment _guarantees_ a 20us uncertainty on what the DMA did.
> 
> The code we have doesn't seem to be broken, so it it ain't broke don't fix it?

Thanks for your review, the purpose of the patch is targeted for a Google reported bug #162832609 regarding sample rate test failure.

The added 20us [1] fixed the alsa_conformance_test, but I am not sure why, and after discussion with KarL, with DPIB, for the same reason 20us added to capture, it should also apply to playback as well, input/output has no difference regarding frame boundary.

An alternative is to use LPIB [2] which also fix the alsa_conformance test, based on the test results, think it's not justified to increase the threshold size to 48 [3] to pass the same tests, I can also open a sof issue for better solution from fresh start.

Note:
[1] - https://github.com/thesofproject/linux/commit/18a556135deeb0964afc93033f2e472db7a4a9f1
[2] - https://github.com/thesofproject/linux/commit/f02996d1e1a5ab2dd10ed632865f687088af08ac
[3] - https://chromium-review.googlesource.com/c/chromiumos/platform/audiotest/+/2375921

**Comment by plbossart** [comment]
> > > V2:
> > > 
> > > 1. Updated based on Kar's suggestion to avoid the 20us, it serves the purpose for alsa_complance_test, but not sure whether DPIB has specific advantage when used here.
> > > 2. I can avoid changes to capture if so preferred, seems the code is deliberated done to fix some issue.
> > 
> > 
> > I don't see the point of using LPIB. This was meant to be used when we transmit on a link, but here these are bursty DMA transfers unrelated to the HDaudio framing. In fact the sample alignment _guarantees_ a 20us uncertainty on what the DMA did.
> > The code we have doesn't seem to be broken, so it it ain't broke don't fix it?
> 
> Thanks for your review, the purpose of the patch is targeted for a Google reported bug #162832609 regarding sample rate test failure.
> 
> The added 20us fixed the alsa_conformance_test, but I am not sure why, and after discussion with KarL, with DPIB, for the same reason 20us added to capture, it should also apply to playback as well, input/output has no difference regarding frame boundary.
> 
> An alternative is to use LPIB which also fix the alsa_conformance test, based on the test results, think it's not justified to increase the threshold size to 48 to pass the same tests, I can also open a sof issue for better solution from fresh start.

Sorry, I don't understand. You provided a fix for the tests, isn't that enough?

**Comment by yongzhi1** [comment]
> Sorry, I don't understand. You provided a fix for the tests, isn't that enough?

The fix I am referring to is the v1 https://github.com/thesofproject/linux/commit/18a556135deeb0964afc93033f2e472db7a4a9f1. I can update the commit message if you approve. Thanks!!

**Comment by plbossart** [comment]
> > Sorry, I don't understand. You provided a fix for the tests, isn't that enough?
> 
> The fix I am referring to is the v1 [18a5561](https://github.com/thesofproject/linux/commit/18a556135deeb0964afc93033f2e472db7a4a9f1). I can update the commit message if you approve. Thanks!!

I really don't get what you are trying to do.

A) you fixed the test - that's independent from SOF drivers.
B) you want us to use LPIB - that impact SOF drivers.

is B) required? That's a massively invasive change and I am not hot on doing this at all. This has been working for SKL+ for many years, we need to be very very careful here. Errors in position reports are the number one cause of problems for audio drivers....

**Comment by yongzhi1** [comment]
> > > Sorry, I don't understand. You provided a fix for the tests, isn't that enough?
> > 
> > 
> > The fix I am referring to is the v1 [18a5561](https://github.com/thesofproject/linux/commit/18a556135deeb0964afc93033f2e472db7a4a9f1). I can update the commit message if you approve. Thanks!!
> 
> I really don't get what you are trying to do.
> 
> A) you fixed the test - that's independent from SOF drivers.
> B) you want us to use LPIB - that impact SOF drivers.
> 
> is B) required? That's a massively invasive change and I am not hot on doing this at all. This has been working for SKL+ for many years, we need to be very very careful here. Errors in position reports are the number one cause of problems for audio drivers....

Hi, @plbossart ,

Please accept my apology if the confusion was caused my description.

Problem statement: the alsa_comformance_test failed for rates test.

With either of the following two patches, we can pass above test with success:
A. https://github.com/thesofproject/linux/commit/18a556135deeb0964afc93033f2e472db7a4a9f1
B. https://github.com/thesofproject/linux/commit/f02996d1e1a5ab2dd10ed632865f687088af08ac

LPIB was also suggested by KarL, and I compared the values for both DPIB and LPIB for spk playback and other test cases, but let's drop B from the discussion for the reason you mentioned.

So in summary, A is the patch for you to review, all other info is try to explain and understand the test itself, thanks!!

**Comment by plbossart** [comment]
> Problem statement: the alsa_comformance_test failed for rates test.
> 
> With either of the following two patches, we can pass above test with success:
> A. [18a5561](https://github.com/thesofproject/linux/commit/18a556135deeb0964afc93033f2e472db7a4a9f1)
> B. [f02996d](https://github.com/thesofproject/linux/commit/f02996d1e1a5ab2dd10ed632865f687088af08ac)
> 
> LPIB was also suggested by KarL, and I compared the values for both DPIB and LPIB for spk playback and other test cases, but let's drop B from the discussion for the reason you mentioned.
> 
> So in summary, A is the patch for you to review, all other info is try to explain and understand the test itself, thanks!!

What about the fix you mentioned above https://chromium-review.googlesource.com/c/chromiumos/platform/audiotest/+/2375921

I really don't want any driver change unless we can prove that the solution is broken, and then an entire revalidation to show that does not impact non-Chrome platforms. We just had a similar issue with your colleagues on BSW-Cyan, and it ran in circles.

**Comment by yongzhi1** [comment]
> > Problem statement: the alsa_comformance_test failed for rates test.
> > With either of the following two patches, we can pass above test with success:
> > A. [18a5561](https://github.com/thesofproject/linux/commit/18a556135deeb0964afc93033f2e472db7a4a9f1)
> > B. [f02996d](https://github.com/thesofproject/linux/commit/f02996d1e1a5ab2dd10ed632865f687088af08ac)
> > LPIB was also suggested by KarL, and I compared the values for both DPIB and LPIB for spk playback and other test cases, but let's drop B from the discussion for the reason you mentioned.
> > So in summary, A is the patch for you to review, all other info is try to explain and understand the test itself, thanks!!
> 
> What about the fix you mentioned above https://chromium-review.googlesource.com/c/chromiumos/platform/audiotest/+/2375921
> 
> I really don't want any driver change unless we can prove that the solution is broken, and then an entire revalidation to show that does not impact non-Chrome platforms. We just had a similar issue with your colleagues on BSW-Cyan, and it ran in circles.

Hi @plbossart , I totally agree and appreciate your concern and caution, the patch 2375921 is a change of "rule of the game", we can change the merge threshold size to give a pass to all the tests (https://github.com/thesofproject/linux/pull/2381#issuecomment-680137349), given the results of both A and B, there is no strong justification to change the test code, at least I am not convinced merge window is the root cause here.

**Comment by plbossart** [comment]
@yongzhi1 I am also nervous because the problem statement is far from clear 

````
The test failed randomly for either rate or mean squared residuals on JSL+
platform, repro rate close to 100%, and with 20us applied, the same test
pass consistently on Waddledoo and Drawlat
````

Your suggested fixes impacts ALL platforms, including the ones that work fine. So there's something odd here and maybe a platform-specific issue. We need to be really careful here.

**Comment by yongzhi1** [comment]
> @yongzhi1 I am also nervous because the problem statement is far from clear
> 
> ```
> The test failed randomly for either rate or mean squared residuals on JSL+
> platform, repro rate close to 100%, and with 20us applied, the same test
> pass consistently on Waddledoo and Drawlat
> ```
> 
> Your suggested fixes impacts ALL platforms, including the ones that work fine. So there's something odd here and maybe a platform-specific issue. We need to be really careful here.

The line of reasoning goes:
1. DPIB read for capture stream is regulated on 20us frame boundary in the current code.
2. frame boundary is the same for both input and output stream (confirmed with KarL)
3. DPIB read for playback stream should be regulated on 20us as well

Although there was no problem reported, I tend to think the code is flawed without 20us on playback stream, sorry to be blunt.

**Comment by kv2019i** [comment]
@yongzhi1 I do share Pierre's concern. The assumptions the conformance test makes do not seem correct. The dump you provided from a failing run on JSL, do not show any invalid values returned from driver. So it would seem to be more about whether it is ok to report accurate values for hw_avail (that might be from a middle of a fast DMA burst), or is driver expected to filter availability values to imitate a burst-free transfer pattern. The "regulating on a 20usec frame boundary" seems to move the implementation towards latter behaviour, but I'm not sure whether this is what we want.

**Comment by yongzhi1** [comment]
> @yongzhi1 I do share Pierre's concern. The assumptions the conformance test makes do not seem correct. The dump you provided from a failing run on JSL, do not show any invalid values returned from driver. So it would seem to be more about whether it is ok to report accurate values for hw_avail (that might be from a middle of a fast DMA burst), or is driver expected to filter availability values to imitate a burst-free transfer pattern. The "regulating on a 20usec frame boundary" seems to move the implementation towards latter behaviour, but I'm not sure whether this is what we want.

Hi, @kv2019i , I respect your expert opinion on the "values returned from driver", and based on your analysis, I can now afford to take a more relaxed attitude towards the test results :)

The counter argument to change test criteria, i.e. using fixed merge_threshold_sz based on burst size or any other way to pre-process the data is that: 1) the same test can be consistently passed for headset capture on my DUT, does the DMA burst transaction only happens for playback? Plus 2) why similar pattern not observed on other platforms to question the test?

localhost ~ #  alsa_conformance_test.py **-C** hw:0,1 --test-suites test_all_pairs 
3 passed, 0 failed
Device Information
	Name: hw:0,1
	Card: sofda7219max983 [sof-da7219max98360a]
	Device: Headset (*) []
	Stream: CAPTURE
	Format: ['S16_LE', 'S24_LE', 'S32_LE']
	Channels: [2]
	Rate: [48000]
	Period_size range: [24, 4080]
	Buffer_size range: [48, 16368]
Test All Pairs
	Set channels 2, format S16_LE, rate 48000: pass
	Set channels 2, format S24_LE, rate 48000: pass
	Set channels 2, format S32_LE, rate 48000: pass
localhost ~ #  alsa_conformance_test.py **-P** hw:0,1 --test-suites test_all_pairs 
1 passed, 2 failed
Device Information
	Name: hw:0,1
	Card: sofda7219max983 [sof-da7219max98360a]
	Device: Headset (*) []
	Stream: PLAYBACK
	Format: ['S16_LE', 'S24_LE', 'S32_LE']
	Channels: [2]
	Rate: [48000]
	Period_size range: [24, 4080]
	Buffer_size range: [48, 16368]
Test All Pairs
	Set channels 2, format S16_LE, rate 48000: pass
	Set channels 2, format S24_LE, rate 48000: fail - Rate error 11.513363 > threshold 10
	Set channels 2, format S32_LE, rate 48000: fail - Rate error 12.266780000000001 > threshold 10

**Comment by plbossart** [comment]
@yongzhi1 the merge_threshold_t value should really be aligned with the DMA burst, if you only look at samples consumed. The default value is 0.1ms, which means an accuracy of about 4.8 samples. The transfers would need to be 10x faster than real-time if we wanted to hide them completely. I think the value is probably too aggressive for such bursty DMA uses. Out of curiosity, would moving to 0.2ms work?

Note that by adding the usleep in the kernel, you also tweak the tests without necessarily improving the precision. The samples consumed will be reported with a delay to userspace, so that would reduce the 'bursty' nature of the DMA.

I can't tell you why it works on other platforms as well as capture, but keep in mind that the ring buffer is in external memory so there are lots of variables we don't control in terms of latency.

**Comment by yongzhi1** [comment]
> @yongzhi1 the merge_threshold_t value should really be aligned with the DMA burst, if you only look at samples consumed. The default value is 0.1ms, which means an accuracy of about 4.8 samples. The transfers would need to be 10x faster than real-time if we wanted to hide them completely. I think the value is probably too aggressive for such bursty DMA uses. Out of curiosity, would moving to 0.2ms work?
> 
> Note that by adding the usleep in the kernel, you also tweak the tests without necessarily improving the precision. The samples consumed will be reported with a delay to userspace, so that would reduce the 'bursty' nature of the DMA.
> 
> I can't tell you why it works on other platforms as well as capture, but keep in mind that the ring buffer is in external memory so there are lots of variables we don't control in terms of latency.

Hi, @plbossart , thanks for further explanation, here is the log [alsa_conformance_test-merge_threshold-0.2ms.txt](https://github.com/thesofproject/linux/files/5158896/alsa_conformance_test-merge_threshold-0.2ms.txt) with --merge_threshold=0.0002.
Maybe the added 20us only contributes to the value averaging than precision as you suggested, @udaymb told me similar test failure also observed on Hatch/helios, and I also see both playback and capture failed for the rate test on Zephyr based SOF on yorp, I agree at this point more time is needed to conclude rather than accepting this patch based on single test on JSL+.

**Comment by plbossart** [comment]
> Hi, @plbossart , thanks for further explanation, here is the log [alsa_conformance_test-merge_threshold-0.2ms.txt](https://github.com/thesofproject/linux/files/5158896/alsa_conformance_test-merge_threshold-0.2ms.txt) with --merge_threshold=0.0002.

Do you mind translating for me what the results are? Pass/Fail?

**Comment by yongzhi1** [comment]
> > Hi, @plbossart , thanks for further explanation, here is the log [alsa_conformance_test-merge_threshold-0.2ms.txt](https://github.com/thesofproject/linux/files/5158896/alsa_conformance_test-merge_threshold-0.2ms.txt) with --merge_threshold=0.0002.
> 
> Do you mind translating for me what the results are? Pass/Fail?

alsa_conformance_test.py does not support --merge_threshold, so I used alsa_conformance_test.

Here are the two pass criteria:
RATE_ERR_CRITERIA default is 10
RATE_CRITERIA_DIFF_PCT default is 0.01, pass range is [47995.2, 48004.8] for rate 48000

Back to the test results:

----------RUN RESULT----------
number of recorders: 1
number of points: 1121
step average: 42.557143
step min: 6
step max: 48
step median: 48
step standard deviation: 11.134099
**rate: 48000.382630 <= this is OK, within 0.01 of 48000**
**rate error: 10.589389 <= this is greater than 10, will result in "Rate error 10.589389  > threshold 10"**
number of underrun: 0
number of overrun: 0

**Comment by plbossart** [comment]
> > > Hi, @plbossart , thanks for further explanation, here is the log [alsa_conformance_test-merge_threshold-0.2ms.txt](https://github.com/thesofproject/linux/files/5158896/alsa_conformance_test-merge_threshold-0.2ms.txt) with --merge_threshold=0.0002.

> number of recorders: 1
> number of points: 1121
> step average: 42.557143
> step min: 6
> step max: 48
> step median: 48
> step standard deviation: 11.134099
> **rate: 48000.382630 <= this is OK, within 0.01 of 48000**
> **rate error: 10.589389 <= this is greater than 10, will result in "Rate error 10.589389 > threshold 10"**

what is 10? we talked about 0.0002 above?

**Comment by yongzhi1** [comment]
> > **rate: 48000.382630 <= this is OK, within 0.01 of 48000**
> > **rate error: 10.589389 <= this is greater than 10, will result in "Rate error 10.589389 > threshold 10"**
> 
> what is 10? we talked about 0.0002 above?

You want to try the test with merge_threshold_t set to 0.2ms instead of current default 0.1ms, right?
10 is RATE_ERR_CRITERIA(also called error "threshold"), the formula to calculate "rate error" was explained here:
https://chromium-review.googlesource.com/c/chromiumos/platform/audiotest/+/2372962/1/alsa_conformance_test/alsa_conformance_recorder.c#209

**Comment by plbossart** [comment]
> > > **rate: 48000.382630 <= this is OK, within 0.01 of 48000**
> > > **rate error: 10.589389 <= this is greater than 10, will result in "Rate error 10.589389 > threshold 10"**
> > 
> > 
> > what is 10? we talked about 0.0002 above?
> 
> You want to try the test with merge_threshold_t set to 0.2ms instead of current default 0.1ms, right?
> 10 is RATE_ERR_CRITERIA(also called error "threshold"), the formula to calculate "rate error" was explained here:
> https://chromium-review.googlesource.com/c/chromiumos/platform/audiotest/+/2372962/1/alsa_conformance_test/alsa_conformance_recorder.c#209

You're losing me, man. to quote your own words, I want the results "with --merge_threshold=0.0002"

**Comment by yongzhi1** [comment]
> You're losing me, man. to quote your own words, I want the results "with --merge_threshold=0.0002"

In the file I attached, I run the same tests multiple times, for each iteration:
1) cmd line 
localhost ~ # alsa_conformance_test -P hw:0,0 -f S24_LE --debug --merge_threshold=0.0002

2) results: search for "rate" and "rate error" for each "RUN RESULT", for example: 

----------RUN RESULT----------
number of recorders: 1
number of points: 1110
step average: 42.979261
step min: 3
step max: 48
step median: 48
step standard deviation: 10.814383
**rate: 48000.134161
rate error: 10.093156**
number of underrun: 0
number of overrun: 0

**Comment by plbossart** [comment]
> **rate error: 10.093156**

What does this mean? are you saying that the merge threshold of 0.0002 is converted into 10?
How are we supposed to interpret your tests?

**Comment by yongzhi1** [comment]
> > **rate error: 10.093156**
> 
> What does this mean? are you saying that the merge threshold of 0.0002 is converted into 10?
> How are we supposed to interpret your tests?

The follow patch was posted for this purpose:
https://chromium-review.googlesource.com/c/chromiumos/platform/audiotest/+/2372962

**Comment by plbossart** [comment]
> > > **rate error: 10.093156**
> > 
> > 
> > What does this mean? are you saying that the merge threshold of 0.0002 is converted into 10?
> > How are we supposed to interpret your tests?
> 
> The follow patch was posted for this purpose:
> https://chromium-review.googlesource.com/c/chromiumos/platform/audiotest/+/2372962

and for that reason, I'm out.
Sorry, I was really expecting simpler answers, you are drowning me with details and I don't have the time or energy to follow-up. You'll have to convince another reviewer.

**Comment by yongzhi1** [comment]
Thanks for all the reviewers and your valuable inputs, although the patch works for jsl+, from the comments I received:
1. The data in the log "do not show any invalid values returned from driver."
2. The merge_threshold_t value should really be aligned with the DMA burst.
3. By adding the usleep in the kernel, you also tweak the tests without necessarily improving the precision.

So looks quite likely the 20us only helps to pass the tests by accident or averaging the burst, so let me drop this PR, thanks for your time!!
